zerver/lib/message.py

```diff
@@ -1317,12 +1326,21 @@ def apply_unread_message_event(
             topic=topic,
         )
 
-        if (
-            stream_id not in state["muted_stream_ids"]
-            # This next check hits the database.
-            and not topic_has_visibility_policy(
-                user_profile, stream_id, topic, UserTopic.VisibilityPolicy.MUTED
+        stream_muted = stream_id in state["muted_stream_ids"]
+        try:
+            user_topic = UserTopic.objects.get(
+                user_profile=user_profile, stream_id=stream_id, topic_name__iexact=topic
             )
+            visibility_policy = user_topic.visibility_policy
+        except UserTopic.DoesNotExist:
+            visibility_policy = UserTopic.VisibilityPolicy.INHERIT
```

Comment:
This bit should probably be extracted as a `get_visibility_policy(user_profile, stream_id, topic)` -- from an abstractions standpoint, we try pretty hard to avoid having lots of copies of queries that need to do `topic_name__iexact` -- we've had bad experiences in the past with mistakes in those, so try to keep that sort of query localized to functions inside the `topic` related modules:

```
$ git grep topic_name__iexact
zerver/lib/stream_subscription.py:                topic_name__iexact=topic_name,
zerver/lib/stream_topic.py:            topic_name__iexact=self.topic_name,
zerver/lib/stream_topic.py:            stream_id=self.stream_id, topic_name__iexact=self.topic_name
zerver/lib/user_topics.py:        topic_name__iexact=topic_name,
zerver/lib/user_topics.py:            user_profile=user_profile, stream_id=stream_id, topic_name__iexact=topic_name
zerver/lib/user_topics.py:        topic_name__iexact=topic_name,
zerver/lib/user_topics.py:        stream_id=stream_id, topic_name__iexact=topic_name
```

Reply:
Ok, abstracted it as 
```python
def get_topic_visibility_policy(
    user_profile: UserProfile,
    stream_id: int,
    topic_name: str,
) -> int:
    try:
        user_topic = UserTopic.objects.get(
            user_profile=user_profile, stream_id=stream_id, topic_name__iexact=topic_name
        )
        visibility_policy = user_topic.visibility_policy
    except UserTopic.DoesNotExist:
        visibility_policy = UserTopic.VisibilityPolicy.INHERIT
    return visibility_policy
```
in `user_topics.py`

We can use this function at a couple of other places in the "automatic follow/unmute PR". I will rebase + update that PR once this gets merged.

Reply:
This is great, sounds good!